### PR TITLE
Add conditional breakpoints

### PR DESCRIPTION
**Pull request by jasonLaster**
Associated Issue: #101

### Summary of Changes

I paired with Giorgio @ppold on this feature last night. This PR is an initial conditional breakpoints implementation, it satisfies the basic requirements and will be polished in subsequent PRs.

User steps:
* click to add a breakpoint
* cmd+click to edit breakpoint
* add a condition
* enter to save

Next steps:
* clicking on the breakpoint while the panel is open should close the panel and remove the breakpoint
* the panel input should be focused when the panel is opened
* style the panel. Perhaps move the description to input placeholder text (shorten dramatically)
* perhaps add a close button
* perhaps update the breakpoint in the breakpoint list to either show it's a conditional or show the condition. Chrome does not do anything here...

### CODE REVIEW

File: public/js/components/Editor.js [hunk 1]
```diff
@@ -26,6 +26,41 @@ function isTextForSource(sourceText) {
...
+    if (e.key != "Enter") {
+      return;
+    }
+
+    setBreakpointCondition(location, e.target.value);
+    closePanel();
+  }
+
+  let panel = document.createElement("div");
+  ReactDOM.render(
```
Comment (jasonLaster) [inline public/js/components/Editor.js]: we likely can replace the react rendering with a simple vanilla approach, but on the other hand this *does* work.

File: public/js/components/Editor.js [hunk 2]
```diff
@@ -66,6 +117,8 @@ const Editor = React.createClass({
         line: line + 1
       });
     } else {
+
+
```
Comment (jasonLaster) [inline public/js/components/Editor.js]: unnecessary whitespace

File: public/js/clients/firefox/commands.js [hunk 3]
```diff
@@ -54,22 +54,26 @@ function setBreakpoint(location, condition, noSliding) {
...
-      id: bpClient.actor,
-      actualLocation: Location(actualLocation)
-    });
+  }).then((res) => onNewBreakpoint(location, res))
+}
+
+function onNewBreakpoint(location, res) {
+  const bpClient = res[1];
+  let actualLocation = res[0].actualLocation;
+  bpClients[bpClient.actor] = bpClient;
```
Comment (jasonLaster) [inline public/js/clients/firefox/commands.js]: we likely can clean this up with some destructuring

Comment (jasonLaster) [inline public/js/clients/firefox/commands.js]: also, setCondition does not return `actualCondition`, but that is likely okay because you're just editing the bp.

File: public/js/actions/breakpoints.js [hunk 4]
```diff
@@ -187,28 +187,31 @@ function toggleAllBreakpoints(shouldDisableBreakpoints: Boolean) {
...
+
+  return ({ dispatch, getState, client }) => {
+    const bp = getBreakpoint(getState(), location);
+    if (!bp) {
+      throw new Error("Breakpoint does not exist at the specified location");
+    }
+    if (bp.loading) {
+      // TODO(jwl): when this function is called, make sure the action
+      // creator waits for the breakpoint to exist
+      throw new Error("breakpoint must be saved");
```
Comment (jasonLaster) [inline public/js/actions/breakpoints.js]: We probably should replace the Errors with `assert` functions

File: public/js/actions/breakpoints.js [hunk 5]
```diff
@@ -187,28 +187,31 @@ function toggleAllBreakpoints(shouldDisableBreakpoints: Boolean) {
...
-  //   return dispatch({
-  //     type: constants.SET_BREAKPOINT_CONDITION,
-  //     breakpoint: bp,
-  //     condition: condition,
-  //     [PROMISE]: Task.spawn(function* () {
-  //       yield client.setBreakpointCondition(bp.get("id"), condition);
-  //     })
-  //   });
-  // };
+  // throw new Error("not implemented");
```
Comment (jasonLaster) [inline public/js/actions/breakpoints.js]: should remove this

File: config/development.json [hunk 6]
```diff
@@ -11,7 +11,8 @@
     "sourceMaps": true,
     "prettyPrint": false,
     "watchExpressions": false,
-    "search": true
+    "search": true,
+    "conditionalBreakpoints": true
```
Comment (jlongster) [inline config/development.json]: I don't think we need to put this behind a flag. Let's just land features and make sure they work from the start. If there are major bugs I think we should fix in the the first PR. But I doubt there are major bugs, there might just be UX improvements we can do. That's the point of landing in nightly though; we have time until uplift to tweak it to get it great.

File: public/js/actions/breakpoints.js (hunk 4 above)
Comment (jlongster) [inline public/js/actions/breakpoints.js]: There no assertions here though; it should always just throw an error. `assert` may evolve in the future to only happen in dev or somethings. But this should always through.

File: public/js/actions/breakpoints.js (hunk 5 above)
Comment (jlongster) [inline public/js/actions/breakpoints.js]: Yeah, go ahead and remove these comments

File: public/js/clients/firefox/commands.js [hunk 7]
```diff
@@ -79,6 +83,12 @@ function removeBreakpoint(breakpointId) {
   return bpClient.remove();
 }
 
+function setBreakpointCondition(breakpointId, location, condition, noSliding) {
+  const bpClient = bpClients[breakpointId];
+  return bpClient.setCondition(threadClient, condition, noSliding)
+    .then((res) => onNewBreakpoint(location, [{}, res]))
```
Comment (jlongster) [inline public/js/clients/firefox/commands.js]: You should probably remove the old breakpoint client from `bpClients` here before we add a new one.

Comment (jlongster) [inline public/js/clients/firefox/commands.js]: Also, can you rename `res` to `bpClient` or something to make it clear that it's just the breakpoint client?

File: public/js/components/Editor.js [hunk 8]
```diff
@@ -26,6 +26,41 @@ function isTextForSource(sourceText) {
...
+      className: "conditional-breakpoint-panel"
+    },
+    dom.div(
+      { className: "header" },
+      `This breakpoint will stop execution only
+       if the following expression is true`
+    ),
+    dom.input({
+      defaultValue: condition,
+      onKeyPress: (e) => onKey(e),
```
Comment (jlongster) [inline public/js/components/Editor.js]: nit: no space around single arg arrow functions

**Comment by jlongster** [comment]
The approach looks fine. A few nits, and I think we should just land it without a flag so that it just gets used and we'll fix bugs when they happen. It doesn't really affect the UI because you only get to it via a key command.

File: public/js/components/Editor.js (hunk 8 above)
Comment (jasonLaster) [inline public/js/components/Editor.js]: we can do better and just do `onKeyPress: onKey` :P

**Comment by jasonLaster** [comment]
I'd like to leave the flag for now. I'm finding some other UX issues to tweak, which can be done in subsequent issues.

ping @jlongster - mind taking another look.
